PluralKit.Bot/Commands/Message.cs

```diff
@@ -21,24 +23,98 @@ public class ProxiedMessage
 {
     private static readonly Duration EditTimeout = Duration.FromMinutes(10);
 
-    // private readonly IDiscordCache _cache;
+    private readonly IDiscordCache _cache;
+    private readonly ModelRepository _repo;
     private readonly IClock _clock;
+    private readonly IMetrics _metrics;
 
     private readonly EmbedService _embeds;
     private readonly LogChannelService _logChannel;
     private readonly DiscordApiClient _rest;
     private readonly WebhookExecutorService _webhookExecutor;
 
     public ProxiedMessage(EmbedService embeds, IClock clock,
-                          DiscordApiClient rest,
+                          DiscordApiClient rest, IMetrics metrics, ModelRepository repo,
                           WebhookExecutorService webhookExecutor, LogChannelService logChannel, IDiscordCache cache)
     {
         _embeds = embeds;
         _clock = clock;
         _rest = rest;
         _webhookExecutor = webhookExecutor;
+        _repo = repo;
         _logChannel = logChannel;
-        // _cache = cache;
+        _cache = cache;
+        _metrics = metrics;
+    }
+
+    public async Task ReproxyMessage(Context ctx)
+    {
+        var msg = await GetMessageToEdit(ctx);
+
+        if (ctx.System.Id != msg.System?.Id)
+            throw new PKError("Can't reproxy a message sent by a different system.");
+
+        var originalMsg = await _rest.GetMessageOrNull(msg.Message.Channel, msg.Message.Mid);
+        if (originalMsg == null)
+            throw new PKError("Could not reproxy message.");
+
+        // Get target member ID
+        var target = await ctx.MatchMember();
+        var previousPtr = ctx.Parameters._ptr;
+
+        // Fetch members and get the ProxyMember for `target`
+        List <ProxyMember> members;
+        using (_metrics.Measure.Timer.Time(BotMetrics.ProxyMembersQueryTime))
+            members = (await _repo.GetProxyMembers(ctx.Author.Id, msg.Message.Guild!.Value)).ToList();
+        var match = members.Find(x => x.Id == target.Id);
+        if (match == null)
+            throw new PKError("Could not reproxy message.");
+
+        // Get a MessageContext for the original message
+        MessageContext? msgCtx =
+               await _repo.GetMessageContext(msg.Message.Sender, msg.Message.Guild!.Value, originalMsg.ChannelId);
+
+        try
+        {
+            var messageChannel = await _cache.GetChannel(originalMsg.ChannelId);
+            var rootChannel = await _cache.GetRootChannel(originalMsg.ChannelId);
```

Comment:
you can't use cache here, clusters don't (currently) share cache - see GetMessage